pkg/volume/flexvolume/flexvolume.go

```diff
@@ -241,76 +245,142 @@ func (f flexVolumeMounter) GetAttributes() volume.Attributes {
 // flexVolumeManager is the abstract interface to flex volume ops.
 type flexVolumeManager interface {
 	// Attaches the disk to the kubelet's host machine.
-	attach(mounter *flexVolumeMounter) (string, error)
+	attach(mounter *flexVolumeMounter, volOptions map[string]string) (string, error)
 	// Detaches the disk from the kubelet's host machine.
 	detach(unmounter *flexVolumeUnmounter, dir string) error
 	// Mounts the disk on the Kubelet's host machine.
-	mount(mounter *flexVolumeMounter, mnt, dir string) error
+	mount(mounter *flexVolumeMounter, mnt, dir string, volOptions map[string]string) error
 	// Unmounts the disk from the Kubelet's host machine.
-	unmount(unounter *flexVolumeUnmounter, dir string) error
+	unmount(unmounter *flexVolumeUnmounter, dir string) error
 }
 
 // SetUpAt creates new directory.
 func (f *flexVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
-
-	notmnt, err := f.blockDeviceMounter.IsLikelyNotMountPoint(dir)
-	if err != nil && !os.IsNotExist(err) {
-		glog.Errorf("Cannot validate mountpoint: %s", dir)
+	err := f.Attach(f.spec, "")
```

Comment:
No--Setup/SetupAt/TearDown/TearDownAt should remain since they are required for mounting and part of the Mounter interface which all plugins should implement. The only thing to do with them is to remove the attach calls from them and move those to the Attacher interface methods.

The MountDevice method, part of the Attacher interface, is designed so that volumes that require attachment, can have a two step mount. MountDevice mounts the device to a global mount path. And then Setup bind mounts the global mount path to the pod. It's up to plugin writers to decide exactly what they will do for each call.

So Flex should implement the following calls:

Mounter:
* `Setup()`/`SetUpAt()` calls `<driver executable> mount`
  * Same as before except no implicit attach call

Unmounter: 
* `TearDown()`/`TearDownAt()` calls `<driver executable> unmount`
  * Same as before except no implicit detach call

Attacher:
* `Attach()` calls `<driver executable> attach` (if defined by executable)
  * New interface method calls existing flex executable method
* `WaitForAttach()` calls `<driver executable> waitforattach` (if defined by executable, it should be if attach is defined)
  * New interface method calls new flex executable method
* `GetDeviceMountPath()` calls `<driver executable> getdevicemountpath` (if defined by executable, it should be if attach is defined)
  * New interface method calls new flex executable method
* `MountDevice()` calls `<driver executable> mountdevice` (if defined by executable)
  * New interface method calls new flex executable method

Detacher:
* `UnmountDevice()` calls `<driver executable> unmountdevice` (if defined by executable)
  * New interface method calls new flex executable method
* `Detach()` calls `<driver executable> detach` (if defined by executable)
  * New interface method calls existing flex executable method
* `WaitForDetach()` calls `<driver executable> waitfordetach` (if defined by executable, it should be if attach is defined)